proc_dash/app.py

```diff
@@ -179,5 +248,58 @@ def reset_table(contents, filename):
     raise PreventUpdate
 
 
+@app.callback(
+    [
+        Output("fig-pipeline-status-all-ses", "figure"),
+        Output("fig-pipeline-status-all-ses", "style"),
+    ],
+    Input("upload-data", "contents"),
+    State("upload-data", "filename"),
+    prevent_initial_call=True,
+)
+def generate_overview_status_fig_for_participants(contents, filename):
+    """
+    If new dataset uploaded, generate stacked bar plot of pipeline_complete statuses per session,
+    grouped by pipeline. Provides overview of the number of participants with each status in a given session,
+    per processing pipeline.
+    """
+    if contents is None:
+        raise PreventUpdate
+    data, total_subjects, sessions, upload_error = util.parse_csv_contents(
+        contents=contents, filename=filename
+    )
+    if upload_error is not None:
```

Comment:
Can't comment on the section in `utility.py` now (or easily) because it wasn't edited in this PR, but I don't think this is a good way of handling errors here. 1) You're handling exceptions with try-except deep inside the codebase - in effect treating them like flow control. That's not a great idea. If something requires an exception to be raised, then just raise it. If a very specific exception should be caught, you can still catch it here with a try-except. But you shouldn't use a catch-all because you will catch even unintended exceptions that you may not be yet aware of. 2) Your specific errors here are user errors - or at least they contain relevant information for the user. Your users won't read the traceback or terminal logs. So you should probably handle an error visually via the UI. https://dash-bootstrap-components.opensource.faculty.ai/docs/components/toast/ is a nice UI element you could have pop up somewhere to alert the user of something that went wrong but was caught.

Reply:
Thanks for the very helpful feedback @surchs ! I've done some refactoring as suggested, but am leaving the migration from error message displaying via a text component to via a dedicated alert element (i.e., a toast) to a separate issue (#30) .